### PR TITLE
[WIP] Receive directly into uninitialized BytesMut memory

### PR DESCRIPTION
**Pull request by Ralith**
Recalling the visibility of `memset` in profiling results, I am attempting here to improve performance by not initializing memory that will immediately be overwritten, and saving a copy from a temporary stack buffer. This works fine, but surprisingly, the impact on the performance of our single benchmark seems to be below the noise floor. I'm not sure we should merge (even trivial) unsafe optimizations without a clear benefit, so this probably needs more work in developing better benchmarks or optimization of the new code.

### CODE REVIEW

**Comment by djc** [comment]
The changes that I had in mind seem to have made it ~10% slower. 😢

File: quinn/src/endpoint.rs [hunk 1]
```diff
@@ -192,16 +191,28 @@ pub(crate) struct EndpointInner {
...
+    recv_buf: BytesMut,
 }
 
 impl EndpointInner {
     fn drive_recv(&mut self, now: Instant) -> Result<bool, io::Error> {
-        let mut buf = [0; 64 * 1024];
         let mut recvd = 0;
         loop {
-            match self.socket.poll_recv(&mut buf) {
+            const MAX_DATAGRAM_SIZE: usize = 64 * 1024;
```
Comment (TimonPost) [inline quinn/src/endpoint.rs]: I would prefer to move constants out of functions.

**Comment by TimonPost** [comment]
Shouldn't this PR be closed? It seems like it is no longer active.

**Comment by Ralith** [comment]
There's nothing wrong with a PR existing for discussion/future reference.

**Comment by Ralith** [comment]
Closing since #548 addressed the worst of this without any unsafety. It'd be nice to get the copies down further, but they don't look like a bottleneck in their current form.

**Pull request closed without merging**